File: platform-mobile/src/main/java/ru/pocketbyte/locolaser/platform/mobile/resource/file/IosSwiftResourceFile.java
/*
 * Copyright © 2017 Denis Shurygin. All rights reserved.
 * Licensed under the Apache License, Version 2.0
 */

package ru.pocketbyte.locolaser.platform.mobile.resource.file;

import org.apache.commons.lang3.text.WordUtils;
import ru.pocketbyte.locolaser.config.WritingConfig;
import ru.pocketbyte.locolaser.platform.mobile.utils.TemplateStr;
import ru.pocketbyte.locolaser.resource.PlatformResources;
import ru.pocketbyte.locolaser.resource.entity.*;
import ru.pocketbyte.locolaser.resource.file.BaseClassResourceFile;
import ru.pocketbyte.locolaser.resource.file.ResourceStreamFile;
import ru.pocketbyte.locolaser.utils.TextUtils;

import java.io.File;
import java.io.IOException;
import java.util.HashSet;
import java.util.Set;

/**
 * Swift class file generator. For generation used Swift version 3.0.
 */
public class IosSwiftResourceFile extends BaseClassResourceFile {

    private static String CLASS_HEADER_TEMPLATE = "import Foundation\r\n"
                                                + "\r\n"
                                                + "public class %s {\r\n";

    private static String CLASS_FOOTER_TEMPLATE = "}";

    private static String PROPERTY_TEMPLATE = "    public static var %s : String {\r\n"
                                            + "        get {\r\n"
                                            + "            return NSLocalizedString(\"%s\", tableName:\"%s\"," +
                                                            " bundle:Bundle.main, value:\"%s\", comment: \"%s\")\r\n"
                                            + "        }\r\n"
                                            + "    }\r\n";

    private static int MAX_LINE_SIZE = 120;

    private String mClassName;
    private String mTableName;

    public IosSwiftResourceFile(File file, String className, String tableName) {
        super(file);
        mClassName = className;
        mTableName = tableName;
    }

    @Override
    public ResMap read() {
        return null;
    }

    @Override
    protected void writeHeaderComment() throws IOException {
        writeStringLn(TemplateStr.GENERATED_CLASS_COMMENT);
    }

    @Override
    protected void writeClassHeader() throws IOException {
        writeStringLn(String.format(CLASS_HEADER_TEMPLATE, mClassName));
    }

    @Override
    protected void writeComment(String comment) throws IOException {
        String commentLinePrefix = "    /// ";
        writeString(commentLinePrefix);
        writeStringLn(WordUtils.wrap(comment, MAX_LINE_SIZE - commentLinePrefix.length()
                , "\r\n" + commentLinePrefix, true));
    }

    @Override
    protected void writeProperty(String propertyName, ResItem item) throws IOException {
        ResValue valueOther = item.valueForQuantity(Quantity.OTHER);

        String comment = valueOther != null && valueOther.comment != null ? valueOther.comment : "";

        writeStringLn(String.format(PROPERTY_TEMPLATE, propertyName, item.key, mTableName,
                valueOther != null ? valueOther.value : null, comment));
    }

    @Override
    protected void writeClassFooter() throws IOException {
        writeString(CLASS_FOOTER_TEMPLATE);
    }

}

File: resource-mobile/src/main/java/ru/pocketbyte/locolaser/mobile/resource/file/IosSwiftResourceFile.kt
/*
 * Copyright © 2017 Denis Shurygin. All rights reserved.
 * Licensed under the Apache License, Version 2.0
 */

package ru.pocketbyte.locolaser.mobile.resource.file

import org.apache.commons.lang3.text.WordUtils
import ru.pocketbyte.locolaser.config.ExtraParams
import ru.pocketbyte.locolaser.mobile.utils.TemplateStr
import ru.pocketbyte.locolaser.resource.entity.*
import ru.pocketbyte.locolaser.resource.file.BaseClassResourceFile

import java.io.File
import java.io.IOException

/**
 * Swift class file generator. For generation used Swift version 3.0.
 */
class IosSwiftResourceFile(
        file: File,
        private val mClassName: String,
        private val mTableName: String
) : BaseClassResourceFile(file) {

    companion object {

        private const val CLASS_HEADER_TEMPLATE = ("import Foundation\r\n"
                + "\r\n"
                + "public class %s {\r\n")

        private const val CLASS_FOOTER_TEMPLATE = "}"

        private const val PROPERTY_TEMPLATE = ("    public static var %s : String {\r\n"
                + "        get {\r\n"
                + "            return NSLocalizedString(\"%s\", tableName:\"%s\"," +
                " bundle:Bundle.main, value:\"%s\", comment: \"%s\")\r\n"
                + "        }\r\n"
                + "    }\r\n")

        private const val MAX_LINE_SIZE = 120

        private fun escapeComment(string: String?): String? {
            return string
                    ?.replace("\r", "\\r")
                    ?.replace("\n", "\\n")
                    ?.replace("\u0009", "  ")
        }

        private fun escapeString(string: String?): String? {
            return string
                    ?.replace("\\", "\\\\")
                    ?.replace("\"", "\\\"")
                    ?.replace("\r", "\\r")
                    ?.replace("\n", "\\n")
                    ?.replace("\u0009", "\\t")
        }
    }

    override fun read(extraParams: ExtraParams?): ResMap? {
        return null
    }

    @Throws(IOException::class)
    override fun writeHeaderComment(resMap: ResMap, extraParams: ExtraParams?) {
        writeStringLn(TemplateStr.GENERATED_CLASS_COMMENT)
    }

    @Throws(IOException::class)
    override fun writeClassHeader(resMap: ResMap, extraParams: ExtraParams?) {
        writeStringLn(String.format(CLASS_HEADER_TEMPLATE, mClassName))
    }

    @Throws(IOException::class)
    override fun writeComment(extraParams: ExtraParams?, comment: String) {
        val commentLinePrefix = "    /// "
        writeString(commentLinePrefix)
        writeStringLn(WordUtils.wrap(escapeComment(comment), MAX_LINE_SIZE - commentLinePrefix.length, "\r\n" + commentLinePrefix, true))
    }

    @Throws(IOException::class)
    override fun writeProperty(extraParams: ExtraParams?, propertyName: String, item: ResItem) {
        val valueOther = item.valueForQuantity(Quantity.OTHER)

        val comment = if (valueOther?.comment != null) valueOther.comment else ""

        writeStringLn(String.format(PROPERTY_TEMPLATE, propertyName, item.key, mTableName,
                if (valueOther != null) escapeString(valueOther.value) else null, escapeString(comment)))
    }

    @Throws(IOException::class)
    override fun writeClassFooter(resMap: ResMap, extraParams: ExtraParams?) {
        writeString(CLASS_FOOTER_TEMPLATE)
    }
}
